### PR TITLE
task/WG-246: new CreateMapModal layout w/ core-wrappers

### PR DESCRIPTION
**Pull request by tjgrafft**
## Overview: ##
Refactoring CreateMapModal to use FieldWrapperFormik and other core-wrapper components from tup. 
## PR Status: ##

* [x] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

* [WG-246](https://tacc-main.atlassian.net/browse/WG-246)

## Summary of Changes: ##

- Implemented core-wrapper components into CreateMapModal instead of solely reactstrap.
- Note that my yup validation is still present, even though tup core-wrappers has some validation built-in (required prop), however the required prop only displays UI, but doesn't prevent the user from submitting the form. The combination of tup's required prop and my yup validation fixes this gap. I also kept my custom file name validation that checks for non-letter, number, hyphen, or underscores to prevent file-naming errors when creating projects.
- Removed Wizard wrapper comp (per Nathan's request) as it was causing issues with our tests and we don't plan to use this wrapper comp in the future anyway.

## Testing Steps: ##
1. Run npm ci
2. Run npm run build
3. Run npm run test
4. Run npm run dev
5. Go to localhost:4200
6. Click 'Create Map' button on main menu
7. Compare modal UI to screenshot below

## UI Photos:
Screenshot 2024-02-12 at 11 11 12 AM

## Alternative Layout with less space between fields: ##
Screenshot 2024-02-12 at 3 46 45 PM

### CODE REVIEW

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (nathanfranklin) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:162]: I see the difference when this (`justify-content-center`) is not added. but I'm perplexed why core-component 'Button" doesn't inherently handle the centering. can we follow up with @wesleyboar on this.

**Review by nathanfranklin — APPROVED** [review]
LGTM 👍 

Just the one comment that is more about our use of <Button>; we can just note this and address this later.

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (nathanfranklin) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:162]: https://github.com/TACC-Cloud/hazmapper/blob/master/react/src/core-components/Button/Button.tsx#L78
[3:56](https://tacc-team.slack.com/archives/D064G7JPJ4S/p1708120588929009)
the Button is a c-button (and at least in CEP see https://github.com/TACC/Core-Portal/blob/main/client/src/styles/components/c-button.css#L41-L49), I see a `justify-content: center;` 

but that `justify-content: center;`  is missing in Core-Styles  https://github.com/TACC/Core-Styles/blob/4216132b39358f9b9033be80e931f67a3d134c2e/src/lib/_imports/components/c-button.css
src/lib/_imports/components/c-button.css .    is not there

Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:162]: **This is a new bug in Core-Styles.** Consider fixing this globally for your project, with a comment above the CSS to remove the code when it's fixed in Core-Styles (feel free to add the URL of this comment).

Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:162]: Actually, anyone is free to fix the bug in Core-Styles. Just open a PR.

Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:162]: [The button now has `justify-content: center;`](https://github.com/TACC/Core-Styles/blob/v2.24.0/src/lib/_imports/components/c-button.css#L24), and this is released — [v2.24.0 on GitHub](https://github.com/TACC/Core-Styles/releases/tag/v2.24.0) and [2.24.0 on NPM](https://www.npmjs.com/package/@tacc/core-styles/v/2.24.0). _**Notice**: Tables are redesigned in this release._

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:104]: **I am sorry.** Some global `c-form…` classes will expect this "parent block"[^1] class to exist.

```suggestion
            <Form class="c-form">
```

Should there be a `<FormFormik>` for this? It would be consistent.

[^1]: The naming structure is a CSS community methodology called BEM (Block, Element, Modifier). Documentation: [internal](https://tacc-main.atlassian.net/wiki/x/xghv), [external](https://en.bem.info/methodology/quick-start/). [↩](https://github.com/TACC-Cloud/hazmapper/pull/208/files#user-content-fnref-1-f9ca5be8fcb41acfa982fb44028de2fa)

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:132]: **Outdated class name.**

The `text-primary` is a holdover from Core-Portal. Please do not use it.

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:133]: **There is an alternative that uses existing code.**

Use an input, but make it `readonly` and `disabled`.

```suggestion
                <FormikInput
                  name="save-location-label"
                  label="Save Location"
                  value={`userData?.username`}
                  readOnly
                  disabled
                />
```

save location via readonly disabled input

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:151]: **Naming.**

Is the close button a toggle? I think not, because once you close it, you do not press the same button to "unclose" it. So… `handleClose`?

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (wesleyboar) [inline react/src/components/CreateMapModal/CreateMapModal.module.css:20]: **Outdated classname? Consider deleting it.**

When I search this page, I only see this class name here, and where it was deleted.

File: react/postcss.config.cjs
Comment (wesleyboar) [inline react/postcss.config.cjs:11]: **Concern.**

I think this app is not correctly installing dependencies for PostCSS.

**TL;DR:** Install `postcss-preset-env`, not `postcss-nested`, because you use the former and it uses (thus isntalls, the latter). And install at [v9](https://www.npmjs.com/package/postcss-preset-env?activeTab=versions), not [v7](https://github.com/TACC-Cloud/hazmapper/blob/task/WG-246-wrapper-modal-layout/react/package-lock.json#L3263), in hopes that the warnings are resolved.

<details>

1. `postcss-preset-env` already installs `postcss-nesting`, and one simply need "turn it on" as was being done.
2. If `npm install` / `npm ci` warnings originate from nested dependencies, report to package maintainer(s), so every user benefits.

_P.S. Article ["How to fix NPM dependency hell" @ "Path to resolution"](https://sylhare.github.io/2022/02/09/How-to-fix-npm-dependency-library-hell.html#path-to-resolution) looks pretty good._

</details>

**Review by wesleyboar — CHANGES_REQUESTED** [review]
Some items to consider. Let's figure out what to change now, later, if at all.

**Review by wesleyboar — COMMENTED** [review]
**Minor. Not blocking.**

I notice horizontal scrollbar on modal after:

1. causing field validation errors
2. focusing on a field
3. taking focus out of a field

https://github.com/TACC-Cloud/hazmapper/assets/62723358/d207fdeb-996f-4d01-a495-c89bfed8f370

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.module.css]: I see this is fixed now.

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:59]: I updated an outdated screenshot to reflect the current state of the Hazmapper UI.

Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:59]: I recognized and added a link to [Bootstrap's Input Group feature](https://getbootstrap.com/docs/4.0/components/input-group/). If it easier to adapt `<Formik…>` wrappers to support that, I'll try not to unleash the Wesley who wants Bootstrap to burn in hell, and rationally consider it.

File: react/postcss.config.cjs
Comment (tjgrafft) [reply react/postcss.config.cjs:11]: npm install postcss-preset-env@latest --save-dev

npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: hazmapper@0.0.0
npm WARN Found: postcss-preset-env@7.8.3
npm WARN node_modules/postcss-preset-env
npm WARN   peer postcss-preset-env@"^7.8.3" from @tacc/core-styles@2.24.1
npm WARN   node_modules/@tacc/core-styles
npm WARN     @tacc/core-styles@"^2.24.1" from the root project
npm WARN   1 more (the root project)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer postcss-preset-env@"^7.8.3" from @tacc/core-styles@2.24.1
npm WARN node_modules/@tacc/core-styles
npm WARN   @tacc/core-styles@"^2.24.1" from the root project

npm install
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: hazmapper@0.0.0
npm ERR! Found: postcss-preset-env@9.4.0
npm ERR! node_modules/postcss-preset-env
npm ERR!   dev postcss-preset-env@"^9.4.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer postcss-preset-env@"^7.8.3" from @tacc/core-styles@2.24.1
npm ERR! node_modules/@tacc/core-styles
npm ERR!   @tacc/core-styles@"^2.24.1" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.

I still get this resolve warning and error when trying to npm install with v9.

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:59]: I got the idea from Bootstrap's input group feature, but I know the Form/FormField comp from core-components has an add-on parameter I was using before. I can go back to using the FormField component and use that add-on if you're ok with that instead of FormikInput (at least for that field, I can still use FormikInput for the Name field).

Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:59]: Alternatively I could add an "add-on" prop to the FormikInput wrapper too.

Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:59]: I do like the look of the s-affixed-input-wrapper--suffix ($) in that example. If we can do that with .hazmapper instead of the dollar sign, that would look good I think.

File: react/postcss.config.cjs
Comment (wesleyboar) [reply react/postcss.config.cjs:11]: Well, fart. Okay. Thanks for trying. I can't argue well enough to spend more time on this. It's an architectural "maybe we should do this" thing, not a blocker.

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:59]: **Yes**, the dollar sign can be replaced with ".hazmapper". **Note:** this modal form's specific instance of usage must set `--width` to a value that tucks the suffix into the field.

Screenshot 2024-02-21 at 6 47 40 PM

I'd prefer code be added to our Formik components to support this, but I haven't thought about the workload. I also haven't considered "[I can go back to using the FormField component](https://github.com/TACC-Cloud/hazmapper/pull/208#discussion_r1498212706)", that may be more efficient.

Would you please try to "go back to using the FormField component" if successful, and offer both options to designers in #cep-design? You may use my screenshot for one. _Even if they want to use the Core-Styles one, I wouldn't judge if you used a separate ticket to convert from Bootstrap to Core-Styles for this field._

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:143]: **Yes.** That's an excellent step for now for this PR.

Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:143]: I plan to make a new task for this change, and address it early March.

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:59]: I plan to make a new task for this change, as well, and address it early March.

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx]: Screenshot 2024-02-26 at 9 20 57 AM

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.module.css:61]: Created a task for the March sprint [WG-264](https://tacc-main.atlassian.net/browse/WG-264)

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx]: Thoughts on the checkbox and description being below the label? See screenshot. This styling/layout is directly in the FormikCheck comp now
Screenshot 2024-02-26 at 11 53 57 AM

Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.tsx]: **Good.** With `FormikCheck` in control of any checkbox, styles can be adjusted[^1] at our leisure[^2] (rather than "fix now so copy-paste does not propagate something different than design").

[^1]: Ideally, it looks like `c-form__field has-type-check` does in [c-form--portal.html](https://tacc.utexas.edu/static/ui/components/detail/c-form--portal.html) ([screenshot](https://github.com/TACC-Cloud/hazmapper/assets/62723358/1b80c7fe-843e-4452-b9cb-37808875600d)), but **zero** rush to achieve that, because `@tacc/core-components` will elt us change it in one spot and propagate.[^2]
[^2]: I'll update [TACC/tup-ui](https://github.com/TACC/tup-ui/) with [TACC-Cloud/hazmapper](https://github.com/TACC-Cloud/hazmapper/) improvements for `@tacc/core-components` release.

**Comment by tjgrafft** [comment]
> **Minor. Not blocking.**
> 
> I notice **horizontal** scrollbar on modal after:
> 
> 1. causing field validation errors
> 2. focusing on a field
> 3. taking focus out of a field
> 
>  horz.scrollbar.mov

This is an issue with different screen sizes. I can see it when toggling the phone view. I'll have to write custom css rules to handle smaller screen sizes bc the Custom File Name's width and height look off on those devices

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:88]: @tjgrafft, regarding:
> custom css rules to handle smaller screen sizes

If this minor issue is explored, and successful CSS changes are not forthcoming, consider a wider modal.

| XL Modal | Size Option |
| - | - |
| xl modal | size |

- https://reactstrap.github.io/?path=/docs/components-modal--modal
- https://react-bootstrap.netlify.app/docs/components/modal/#optional-sizes

_P.S. I think all the elements in the modal content should be able to flexible enough to not need media queries._

Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:88]: I took a stab at the .hazmapper suffix while fixing the horizontal scroll and handling the different screen sizes--see screenshot below. Probably not a finished product and needs some polishing, but it's a start. If that code and the UI look acceptable, do you mind being the second approver for this PR? Once it's approved I'll merge it and handle any future tweaks in the March sprint. Thanks!
Screenshot 2024-02-28 at 1 40 47 PM

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.tsx]: I see `package.json` loads core-styles with fix. I resolve this thread. If I have done so in error, please unresolve it.

File: react/src/components/CreateMapModal/CreateMapModal.module.css
Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.module.css]: I just now noticed the work toward this in d184aa1. Resolving conversation.

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (wesleyboar) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:88]: Oh, hey, cool. No problem. I only just confirmed you already handled everything I asked about. Sorry for the delay. Thanks for all the effort.

**Review by wesleyboar — APPROVED** [review]
I have not re-tested. I still approve based on commits since review, and available screenshots of success. This is a great complex form from which other forms can borrow pieces.

File: react/src/components/CreateMapModal/CreateMapModal.tsx
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:88]: All good, thanks for taking the time to review this PR and helping me with the UI for this modal/form. I appreciate it!

**Pull request merged**
